### ISSUE TITLE
Mixed compilation with `.java` sources causes cyclic inheritance error

### ISSUE THREAD

**Issue by WojciechMazur** [lampepfl/dotty#21533]
## Compiler version
All Scala versions (both Scala 3 and Scala 2)

## Minimized code
Important: Compiles correctly when build server is not used (`scala --server=false`), otherwise compilation of java source yields cyclic inheritance error

```scala
// ReporterDefs.scala
class ConsoleReporter extends ConsoleReporter.AbstractConsoleReporter
object ConsoleReporter{
  abstract class AbstractConsoleReporter 
}

abstract class AbstractReporter
```

```Java
// Reporters.java
class Reporters {
  abstract class Works extends AbstractReporter{};
  abstract class Fails1 extends ConsoleReporter{};
  abstract class Fails2 extends ConsoleReporter.AbstractConsoleReporter{};
}
```

## Output

```scala
[error] cyclic inheritance involving ConsoleReporter
[error]   abstract class Fails1 extends ConsoleReporter{};
[error]                                 ^^^^^^^^^^^^^^^
```

## Expectation
Should compile also when build server is used

**Comment by Gedochao** [comment]
@tgodzik any clue?
This looks like some quirk of Bloop, but it'd be weird if we didn't have it reported before 🤔

**Comment by dwijnand** [comment]
Is this reproducible with straight scala-cli?

**Comment by Gedochao** [comment]
> Is this reproducible with straight scala-cli?

Yep. Tested on `scala-cli` 1.5.0 and 1.5.0-17-g00e4c88c1-SNAPSHOT.

**Comment by Gedochao** [comment]
This is not related to the runner.
The only reason the error doesn't pop with `--server=false` is that `--server=false` ignores `.java` files right now.
Also see:
- https://github.com/VirtusLab/scala-cli/issues/3196

Since this happens in Bazel as well, I don't think it's Bloop-related.
We should investigate on the compiler side of things.
